backend/chatbot.py

```diff
@@ -0,0 +1,146 @@
+import spacy
+from environment import shapes, colors, addShape,delShape,findShape,moveShape,holdShape, GridSize, Grid,showGrid
+
+#Create function for conjunctions(and) to make sentences easier to read
+
+
+nlp = spacy.load('en_core_web_sm')
+
+actions = {'find':findShape,'delete':delShape,'add':addShape,'move':moveShape,'pick':holdShape}
+
+#Takes the format "[action] [adj] [noun] [x] [y]"
+def readSentence(sentence):
+
+    x = nlp(sentence)
+
+    color,shape,action = '','',''
+    row,col = -1,-1,
+    for token in x:
+        if token.pos_ == 'VERB':
+            action = token.text.lower()
+        if token.pos_ == 'ADJ':
+            color = token.text.lower()
+        if token.pos_ == 'NOUN':
+            shape = token.text.lower()
+        if token.pos_ == 'NUM':
+            if row == -1:
+                row = int(token.text)
+            else:
+                col = int(token.text)
+
+        #print(token, token.tag_, token.pos_, spacy.explain(token.tag_), token.head.text, token.dep_)
+        #example output "Add VB VERB verb, base form Add ROOT"
+
+    return(shape,color,action,row,col)
+
+
+#Checks to see if the semantics are in our lists:
+#check[noun,adjective,verb,location]
+#0 -> no error, 1 -> error
+def checkSemantics(noun,adj,verb,row,col):
+    check = [0,0,0,0]
+    if noun not in shapes:
+        check[0] = 1
+    if adj not in colors:
+        check[1] = 1
+    if verb not in actions:
+        check[2] = 1
+    if row > GridSize or col > GridSize \
+        or row < 0 or col < 0:
+        check[3] = 1
+
+    return check
```

Comment:
could this return true or false?

Reply:
I was planning to use the function to give a list of flags so it would be easier for condition checking since we are hard coding the decision process

Reply:
dear respected colleague, please separate checks into individual functions. many thanks 👍